### PR TITLE
Modify build.sh and test.sh scripts for ppc64le jenkins build and test

### PR DESCRIPTION
**Pull request by avmgithub**
Initial jenkins builds / test scripts for ppc64le.

### CODE REVIEW

**Comment by ezyang** [comment]
* What's the plan for getting PPC machines connected to Jenkins, with enough capacity so that they can keep up with the volume of builds we request?
* Can we cross-compile for PPC? That will help reduce the number of PPC machines we need for testing.
* Can we avoid copy pasting the scripts?

**Comment by avmgithub** [comment]
@ezyang answer to your questions:
1. There is currently a CI build environment from https://powerci.osuosl.org/ . It is a shared environment for other jobs.  I'm already doing build/test there for pytorch master.  When I was working with @yf225  , he said pytorch PR builds should at least accommodate 4 simultaneous builds (the environment should be able to accommodate that).  I believe @yf225 also already has the hooks from your ci to be able to trigger builds on our CI.   The plan is to start with a single build/test image using Ubuntu CUDA image from nvidia,  CUDA 9.1 , cudnn 7.x , python3 from anaconda.     When the environment is ready,  we can switch to cuda 9.2.   I am still trying to eliminate random environment failures before we can integrate with your CI.  We can talk about it more when I am ready.  For now I hope to get the scripts ready for ppc64le.
2. Cross-compile may work (I've never tried it or set it up with nvcc).   We do have a toolchain cross compiler that we can setup for x86 (I've used this on other non-CUDA projects).  If you're willing to try it,  I can help test the produced binaries.   Just FYI ,  when building pytorch on ppc64le it takes between 7-14 minutes to finish when using ccache (so it does not take a lot of resources).
3. I'll take your suggestion on using your build.sh and test.sh  and ifdef for ppc64le to avoid duplication of code in different files.

**Comment by ezyang** [comment]
So, I looked at your modified patch: you're still doing what's effectively a copy paste, since you have a giant if-block conditioned on ppc, versus not. If you're going to do that, you might as well just have another file.

What I would like to see is a *localized* block that sets the extra environment variables that PPC needs, and no reindenting of the build script. It would be even better if we had a PPC docker image (similar to the docker images we have for our Linux builds), which can have these environment variables pre-set. The definitions for these currently live in https://github.com/pietern/pytorch-dockerfiles

**Comment by avmgithub** [comment]
@ezyang Ok , I got it.  For build.sh I can use it as is except for the line   WERROR=1 python setup.py install .  I am not able to use the line with WERROR=1 .    For test.sh ,   how are you able to run the test while in the pytorch directory?    I have to do a "cd .."  out of the directory.  Otherwise I get an error like  below.    

+ test_python_nn
+ python test/run_test.py --include nn --verbose
Traceback (most recent call last):
  File "test/run_test.py", line 14, in <module>
    import torch
  File "/home/freddie/builder/jenkins/pytorch/pytorch/torch/__init__.py", line 84, in <module>
    from torch._C import *
ModuleNotFoundError: No module named 'torch._C'

this is because there is a torch directory in the pytorch repo.   Does this not happen on x86 ?

**Comment by avmgithub** [comment]
@ezyang, when you get the chance please review the latest revision.

File: .jenkins/pytorch/build.sh [hunk 1]
```diff
@@ -71,7 +71,12 @@ if [[ "$BUILD_ENVIRONMENT" == *trusty-py3.6-gcc5.4* ]]; then
   export DEBUG=1
 fi
 
-WERROR=1 python setup.py install
+if [[ "$BUILD_ENVIRONMENT" == *ppc64le* ]]; then
+  export TORCH_CUDA_ARCH_LIST=6.0
```
Comment (ezyang) [inline .jenkins/pytorch/build.sh]: Put this up with the other TORCH_CUDA_ARCH_LIST. And I guess you can move `WERROR=1` out into an actual export, so you can write it conditionally without copying `python setup.py install`.

File: .jenkins/pytorch/test.sh [hunk 2]
```diff
@@ -80,8 +80,13 @@ test_aten() {
     # NB: the ATen test binaries don't have RPATH set, so it's necessary to
     # put the dynamic libraries somewhere were the dynamic linker can find them.
     # This is a bit of a hack.
-    ln -s "$TORCH_LIB_PATH"/libcaffe2* build/bin
-    ln -s "$TORCH_LIB_PATH"/libnccl* build/bin
+    if [[ "$BUILD_ENVIRONMENT" == *ppc64le* ]]; then
+      SUDO=sudo 
+    fi
```
Comment (ezyang) [inline .jenkins/pytorch/test.sh]: Why is this needed?

**Comment by avmgithub** [comment]
@ezyang when you get the chance , please review.   I tried to do the export WERROR=1  in a separate conditional , but it created a problem when building the ATen install.  So it looks like it needs to be in line with the python setup.py install.

File: .jenkins/pytorch/build.sh [hunk 3]
```diff
@@ -65,13 +65,21 @@ if ([[ "$BUILD_ENVIRONMENT" == *cuda* ]] || [[ "$BUILD_ENVIRONMENT" == *gcc7* ]]
 fi
 
 # Target only our CI GPU machine's CUDA arch to speed up the build
-export TORCH_CUDA_ARCH_LIST=5.2
+export TORCH_CUDA_ARCH_LIST="5.2 6.0"
```
Comment (ezyang) [inline .jenkins/pytorch/build.sh]: Err, we don't want this unconditionally, right? Because the regular CI machines are all 5.2, so the 6.0 build is just a waste of compilation cycles.

File: .jenkins/pytorch/build.sh [hunk 4]
```diff
@@ -65,13 +65,21 @@ if ([[ "$BUILD_ENVIRONMENT" == *cuda* ]] || [[ "$BUILD_ENVIRONMENT" == *gcc7* ]]
...
 
 if [[ "$BUILD_ENVIRONMENT" == *trusty-py3.6-gcc5.4* ]]; then
   export DEBUG=1
 fi
 
-WERROR=1 python setup.py install
+# ppc64le build fails when WERROR=1
+# set only when building other archtectures
```
Comment (ezyang) [inline .jenkins/pytorch/build.sh]: nit: architectures

**Comment by avmgithub** [comment]
@ezyang please review when you get the chance,  hope the changes are OK

**Review by ezyang — APPROVED** [review]
